share/spack/setup-env.sh

```diff
@@ -60,21 +60,62 @@ _spack_shell_wrapper() {
         emulate -L sh
     fi
 
-    # accumulate flags meant for the main spack command
-    # the loop condition is unreadable, but it means:
-    #     while $1 is set (while there are arguments)
-    #       and $1 starts with '-' (and the arguments are flags)
+    # accumulate flags meant for spack itself preceding the command
     _sp_flags=""
-    while [ ! -z ${1+x} ] && [ "${1#-}" != "${1}" ]; do
-        _sp_flags="$_sp_flags $1"
-        shift
+    help_or_version_was_used=0
+    while [ -n "$1" ]; do
+        case "$1" in
+            --color|--config|--sorted-profile|--lines|--env-dir|--env)
+                # long flags taking required values
+                _sp_flags="$_sp_flags $1 $2"
+                shift 2
+                ;;
+            --help|--version)
+                # special early return flags
+                help_or_version_was_used=1
+                _sp_flags="$_sp_flags $1"
+                shift
+                ;;
+            --*)
+                # long flags taking no value
+                _sp_flags="$_sp_flags $1"
+                shift
+                ;;
+            -*)
+                # loop over all single character flags (e.g. -kle[env]), but
+                # skip the dash before the flag, so start at 1
+                for (( i=1 ; i<${#1}; i++ )); do
+                    flag="${1:$i:1}"
+                    case "$flag" in
+                        c|C|e|D)
```

Comment:
I don't think this is accurate.

If the flag is `-Cd`, that's invalid and shouldn't take an argument. It's only the last of the string of short flags that needs to be checked for accepting a value.

I think the case `-*C|-*c|-*e|-*D` covers those cases.

Reply:
Not sure what you mean, `-Cd` and `-C d` are equivalent and valid

Reply:
e.g. this is fine:

```$ spack -dCnonexistent/path spec wut
==> [2021-03-18-18:04:11.170923] ConfigError: config scope is not a directory: 'nonexistent/path'
==> [2021-03-18-18:04:11.171243] Error: config scope is not a directory: 'nonexistent/path'
Traceback (most recent call last):
  File "/home/harmen/spack/lib/spack/spack/main.py", line 745, in main
    setup_main_options(args)
  File "/home/harmen/spack/lib/spack/spack/main.py", line 449, in setup_main_options
    spack.config.set('config:debug', True, scope='command_line')
  File "/home/harmen/spack/lib/spack/spack/config.py", line 894, in set
    return config.set(path, value, scope)
  File "/home/harmen/spack/lib/spack/llnl/util/lang.py", line 582, in __getattr__
    return getattr(self.instance, name)
  File "/home/harmen/spack/lib/spack/llnl/util/lang.py", line 572, in instance
    self._instance = self.factory()
  File "/home/harmen/spack/lib/spack/spack/config.py", line 796, in _config
    _add_command_line_scopes(cfg, command_line_scopes)
  File "/home/harmen/spack/lib/spack/spack/config.py", line 761, in _add_command_line_scopes
    raise ConfigError("config scope is not a directory: '%s'" % path)
spack.config.ConfigError: config scope is not a directory: 'nonexistent/path'
```